File: apps/downloadx/src/chunkScheduler.ts
import type { Chunk } from './chunk.js';

export interface PlanOptions {
  totalSize: number;
  targetChunkCount: number;
  minChunkSize: number;
  /** Bytes already placed on disk per offset-sorted plan index (for resume). */
  resumeFrom?: ReadonlyArray<{ offset: number; length: number; downloadedBytes: number }>;
}

export interface ChunkPlan {
  offset: number;
  length: number;
  downloadedBytes: number;
}

/**
 * Divides a `totalSize` byte file into at most `targetChunkCount` contiguous
 * chunks, each no smaller than `minChunkSize` (except the last, which gets
 * the remainder). Single-chunk fallback when the file is smaller than
 * `minChunkSize`.
 */
export function planChunks(opts: PlanOptions): ChunkPlan[] {
  if (opts.resumeFrom && opts.resumeFrom.length > 0) {
    // Preserve the historical plan on resume so offsets line up with what's
    // already on disk.
    return opts.resumeFrom.map((c) => ({ ...c }));
  }
  if (opts.totalSize <= 0) {
    return [{ offset: 0, length: 0, downloadedBytes: 0 }];
  }
  if (opts.totalSize <= opts.minChunkSize || opts.targetChunkCount <= 1) {
    return [{ offset: 0, length: opts.totalSize, downloadedBytes: 0 }];
  }
  const count = Math.min(
    opts.targetChunkCount,
    Math.max(1, Math.floor(opts.totalSize / opts.minChunkSize)),
  );
  const base = Math.floor(opts.totalSize / count);
  const plans: ChunkPlan[] = [];
  let offset = 0;
  for (let i = 0; i < count; i += 1) {
    const last = i === count - 1;
    const length = last ? opts.totalSize - offset : base;
    plans.push({ offset, length, downloadedBytes: 0 });
    offset += length;
  }
  return plans;
}

export interface SplitCandidate {
  chunk: Chunk;
  /** The slice taken from `chunk`'s tail, to be issued as a new chunk. */
  newRange: { offset: number; length: number };
  reason: 'slow' | 'failed' | 'completed-reassign';
}

export interface FindSplitOptions {
  /** All non-completed chunks currently part of the download. */
  activeChunks: readonly Chunk[];
  /** Upper bound (inclusive) on live chunk count. */
  maxChunks: number;
  /** Minimum bytes we'll leave on either side of a split. */
  minChunkSize: number;
  /** Why we're looking for a split (drives the `reason` field on the result). */
  trigger: 'slow' | 'failed' | 'completed-reassign';
}

/**
 * Picks the chunk best suited to donate its tail to a new worker.
 *
 * Scoring:
 *   1. Prefer `stalled` over `poor` over `good` (slower chunks lose more bytes).
 *   2. Among equal quality tiers, prefer the one with the most remaining bytes
 *      — that's the one we'll shave the most time off by parallelising.
 *
 * Returns `null` when:
 *   - we're already at max chunk count
 *   - no chunk has enough remaining bytes to split
 */
export function findSplitCandidate(opts: FindSplitOptions): SplitCandidate | null {
  if (opts.activeChunks.length >= opts.maxChunks) return null;

  const eligible = opts.activeChunks.filter(
    (c) =>
      c.status === 'downloading' &&
      c.remainingBytes >= opts.minChunkSize * 2,
  );
  if (eligible.length === 0) return null;

  const scored = eligible
    .map((chunk) => ({ chunk, score: scoreChunk(chunk) }))
    .sort((a, b) => b.score - a.score);

  const pick = scored[0];
  if (pick === undefined) return null;

  const newRange = pick.chunk.truncateTail(opts.minChunkSize);
  if (newRange === null) return null;

  return {
    chunk: pick.chunk,
    newRange,
    reason: opts.trigger,
  };
}

function scoreChunk(chunk: Chunk): number {
  const qualityWeight =
    chunk.quality === 'stalled' ? 3 : chunk.quality === 'poor' ? 2 : 1;
  // Remaining bytes normalised to MiB so the quality multiplier dominates
  // for comparable remainders, but size still breaks ties.
  const remainingMiB = chunk.remainingBytes / (1024 * 1024);
  return qualityWeight * 1000 + remainingMiB;
}

File: apps/downloadx_dart/lib/src/chunk_scheduler.dart
import 'chunk.dart';
import 'types.dart';

/// A planned byte range assigned to one chunk before download starts.
class ChunkPlan {
  /// Byte offset from the start of the file.
  final int offset;

  /// Number of bytes this chunk is responsible for.
  final int length;

  /// Bytes already written for this range (non-zero on resume).
  final int downloadedBytes;

  /// Creates a [ChunkPlan] for the given byte range.
  const ChunkPlan({
    required this.offset,
    required this.length,
    required this.downloadedBytes,
  });
}

/// Options for [planChunks].
class PlanOptions {
  /// Total file size in bytes.
  final int totalSize;

  /// Desired number of chunks (may be reduced when file is small).
  final int targetChunkCount;

  /// Minimum bytes per chunk; prevents splitting into too-small pieces.
  final int minChunkSize;

  /// Bytes already placed on disk per offset-sorted plan index (for resume).
  final List<ChunkPlan>? resumeFrom;

  /// Creates a [PlanOptions] with the given parameters.
  const PlanOptions({
    required this.totalSize,
    required this.targetChunkCount,
    required this.minChunkSize,
    this.resumeFrom,
  });
}

/// Divides a [totalSize] byte file into at most `targetChunkCount` contiguous
/// chunks, each no smaller than `minChunkSize` (except the last, which gets the
/// remainder). Single-chunk fallback when the file is smaller than minChunkSize.
List<ChunkPlan> planChunks(PlanOptions opts) {
  final resume = opts.resumeFrom;
  if (resume != null && resume.isNotEmpty) {
    // Preserve the historical plan on resume so offsets line up with disk.
    return resume
        .map((c) => ChunkPlan(
              offset: c.offset,
              length: c.length,
              downloadedBytes: c.downloadedBytes,
            ))
        .toList();
  }
  if (opts.totalSize <= 0) {
    return const [ChunkPlan(offset: 0, length: 0, downloadedBytes: 0)];
  }
  if (opts.totalSize <= opts.minChunkSize || opts.targetChunkCount <= 1) {
    return [ChunkPlan(offset: 0, length: opts.totalSize, downloadedBytes: 0)];
  }
  final count = _min(
    opts.targetChunkCount,
    _max(1, opts.totalSize ~/ opts.minChunkSize),
  );
  final base = opts.totalSize ~/ count;
  final plans = <ChunkPlan>[];
  var offset = 0;
  for (var i = 0; i < count; i += 1) {
    final last = i == count - 1;
    final length = last ? opts.totalSize - offset : base;
    plans.add(ChunkPlan(offset: offset, length: length, downloadedBytes: 0));
    offset += length;
  }
  return plans;
}

/// A chunk selected to donate its tail to a new parallel worker.
class SplitCandidate {
  /// The chunk whose tail will be truncated and reassigned.
  final Chunk chunk;

  /// The slice taken from [chunk]'s tail, to be issued as a new chunk.
  final ByteRange newRange;

  /// Why this split was triggered.
  final SplitReason reason;

  /// Creates a [SplitCandidate].
  const SplitCandidate({
    required this.chunk,
    required this.newRange,
    required this.reason,
  });
}

/// Options for [findSplitCandidate].
class FindSplitOptions {
  /// Currently active chunks to evaluate for splitting.
  final List<Chunk> activeChunks;

  /// Upper bound on total live chunks; no split when already at capacity.
  final int maxChunks;

  /// Minimum bytes a chunk must have remaining to be eligible for splitting.
  final int minChunkSize;

  /// The reason that triggered this split search.
  final SplitReason trigger;

  /// Creates a [FindSplitOptions].
  const FindSplitOptions({
    required this.activeChunks,
    required this.maxChunks,
    required this.minChunkSize,
    required this.trigger,
  });
}

/// Picks the chunk best suited to donate its tail to a new worker.
///
/// Scoring:
///   1. Prefer `stalled` over `poor` over `good` (slower chunks lose more).
///   2. Among equal quality tiers, prefer the one with the most remaining bytes.
///
/// Returns null when already at max chunk count or no chunk has enough
/// remaining bytes to split.
SplitCandidate? findSplitCandidate(FindSplitOptions opts) {
  if (opts.activeChunks.length >= opts.maxChunks) return null;

  final eligible = opts.activeChunks
      .where((c) =>
          c.status == ChunkStatus.downloading &&
          c.remainingBytes >= opts.minChunkSize * 2)
      .toList();
  if (eligible.isEmpty) return null;

  eligible.sort((a, b) => _scoreChunk(b).compareTo(_scoreChunk(a)));
  final pick = eligible.first;

  final newRange = pick.truncateTail(opts.minChunkSize);
  if (newRange == null) return null;

  return SplitCandidate(chunk: pick, newRange: newRange, reason: opts.trigger);
}

double _scoreChunk(Chunk chunk) {
  final qualityWeight = chunk.quality == ChunkQuality.stalled
      ? 3
      : chunk.quality == ChunkQuality.poor
          ? 2
          : 1;
  // Remaining bytes normalised to MiB so the quality multiplier dominates for
  // comparable remainders, but size still breaks ties.
  final remainingMiB = chunk.remainingBytes / (1024 * 1024);
  return qualityWeight * 1000 + remainingMiB;
}

int _min(int a, int b) => a < b ? a : b;
int _max(int a, int b) => a > b ? a : b;
